server/hotrod/src/main/scala/org/infinispan/server/hotrod/ClientEventSenders.scala

```diff
@@ -0,0 +1,392 @@
+package org.infinispan.server.hotrod
+
+import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.atomic.AtomicLong
+
+import io.netty.channel.Channel
+import org.infinispan.container.versioning.NumericVersion
+import org.infinispan.notifications.Listener
+import org.infinispan.notifications.cachelistener.annotation.{CacheEntryCreated, CacheEntryExpired, CacheEntryModified, CacheEntryRemoved}
+import org.infinispan.notifications.cachelistener.event.Event.Type
+import org.infinispan.notifications.cachelistener.event.{CacheEntryCreatedEvent, CacheEntryEvent, CacheEntryModifiedEvent, CacheEntryRemovedEvent}
+import org.infinispan.server.hotrod.ClientListenerRegistry.{CustomRaw, CustomPlain, Plain, ClientEventType}
+import org.infinispan.server.hotrod.Events.{CustomEvent, CustomRawEvent, KeyEvent, KeyWithVersionEvent}
+import org.infinispan.server.hotrod.OperationResponse._
+import org.infinispan.server.hotrod.logging.Log
+
+object ClientEventSenders extends Log {
+
+  val isTrace = isTraceEnabled
+
+  private val messageId = new AtomicLong()
+
+  def createClientSender(includeState: Boolean, ch: Channel, version: Byte,
+            cache: Cache, listenerId: Bytes, eventType: ClientEventType, listenerInterests: Byte): AnyRef = {
+    val compatibility = cache.getCacheConfiguration.compatibility()
+    (includeState, compatibility.enabled()) match {
+      case (false, false) =>
+        createStatelessClientEventSender(ch, listenerId, version, eventType, listenerInterests)
+      case (true, false) =>
+        createStatefulClientEventSender(ch, listenerId, version, eventType, listenerInterests)
+      case (false, true) =>
+        val delegate = new StatelessClientEventSender(ch, listenerId, version, eventType)
+        createStatelessCompatibilityClientEventSender(delegate, HotRodTypeConverter(compatibility.marshaller()), listenerInterests)
+      case (true, true) =>
+        val delegate = new StatelessClientEventSender(ch, listenerId, version, eventType)
+        createStatefulCompatibilityClientEventSender(delegate, HotRodTypeConverter(compatibility.marshaller()), listenerInterests)
+    }
+  }
+
+  def createStatelessClientEventSender(ch: Channel, listenerId: Bytes, version: Byte, eventType: ClientEventType, listenerInterests: Byte) = {
+    listenerInterests match {
```

Comment:
Is this still required to have the various `*Sender` classes?  Can't we just have 1 class with all of the remote notification annotations then apply the filter?